### PR TITLE
mon,osd: do not use crush_device_class file to initalize class for new osds

### PR DESCRIPTION
**Pull request by liewegas**
Instead, pass it to the 'osd new' command.

TODO:
- [ ] update ceph-disk to do this
- [ ] update ceph-volume to do this

### CODE REVIEW

**Comment by liewegas** [comment]
Note that this removes support for the old file, so we need to fix ceph-disk and ceph-volume, too.

**Comment by alfredodeza** [comment]
Isn't it possibly to keep both for now? Once we get to the N release we can remove the support for the file-based class

**Comment by liewegas** [comment]
We could, but I think the file method is problematic.  The intention was to set the class for new osds, but if you do that, then a year later decide you don't want the class (ceph osd crush rm-device-class or whatever), the next time the osd restarts it will get reset to the class specified at creation-time.  If we can change ceph-disk and ceph-volume now to pass it via osd new that would be better.

I don't think it's a version compatibliity issue since the installed ceph-disk and ceph-volume should match the ceph-osd version (i.e. no reason to wait for any particular release; we could even backport this change).

**Comment by andrewschoen** [comment]
This looks good to me. I'm working on the ``ceph-volume`` and ``ceph-disk`` support for this here: https://github.com/ceph/ceph/compare/wip-bz1498521

Do we want to get this merged so I can test it with my branch?

**Comment by andrewschoen** [comment]
@liewegas I've been looking into ``ceph-disk`` support for this and it's more involved than I initially thought. It doesn't look like ``ceph-disk`` calls ``osd new`` at all during the prepare step unless you're using dmcrypt. It only calls ``osd new`` during activate if the ``whoami`` file doesn't exist, but activate doesn't have access to the value for ``--crush-device-class`` unless the ``crush_device_class`` file exists in the osd directory, which kind of defeats the whole purpose. Would you mind taking a look? Maybe I'm missing something here. Perhaps we could leave support for the file until ``ceph-disk`` is fully deprecated and then remove it?

I've got the ``ceph-volume`` support for this ready in https://github.com/ceph/ceph/pull/19949

**Comment by liewegas** [comment]
sounds reasonable.  i can drop the osd change removing the file support.

**Comment by tchaikov** [comment]
```
2018-02-01T10:34:17.523 INFO:tasks.workunit.client.0.mira077.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1269: test_mon_osd_create_destroy:  ceph osd new d298768e-3718-423c-ba92-62d1067a08c9 -i /tmp/cephtool.YZk/secret.npPddl
2018-02-01T10:34:17.523 INFO:tasks.workunit.client.0.mira077.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1269: test_mon_osd_create_destroy:  grep EINVAL
2018-02-01T10:34:17.726 INFO:tasks.workunit.client.0.mira077.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: test_mon_osd_create_destroy:  rm -fr /tmp/cephtool.YZk
```

see http://pulpito.ceph.com/kchai-2018-02-01_09:17:08-rados-wip-kefu-testing-2018-02-01-1258-distro-basic-mira/2137709/

i don't think it's relevant . but this PR is the only one that touches this area. i need to rerun this test locally to investigate this failure further.

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -6890,30 +6917,38 @@ int OSDMonitor::prepare_command_osd_new(
...
+  string device_class;
+  auto p = params.find("crush_device_class");
+  if (p != params.end()) {
+    device_class = p->second;
+    dout(20) << __func__ << " device_class will be " << device_class << dendl;
+  }
   string cephx_secret, lockbox_secret, dmcrypt_key;
   bool has_lockbox = false;
-  bool has_secrets = (!secrets.empty());
+  bool has_secrets = params.count("cephx_secret")
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: this breaks the tests at https://github.com/ceph/ceph/blob/master/qa/workunits/cephtool/test.sh#L1269 . before this change, before this change the `secrets` is a map of `{"asd" => "", "}" => ""}`, so it's not "empty".  and a non-empty secret map without "cephx_secret" is a considered invalid input before this change.

but with this change, we practically ignore the "params" if it does not contain anything interesting. even if it is a map parsed from gibberish.

so my question is: shall we validate the params by checking if it contains some keys we do not understand? something like
```c++
for (auto param& : params) {
  if (param.first == "cephx_secret")) {
    has_secrets = true;
  } else if (param.first == "cephx_lockbox_secret") {
    has_secrets = true;
    has_lockbox_secret = true;
  } else if (param.first == "dmcrypt_key") {
    has_secrets = true;
    has_dmcrypt_key = true;
  } else if (param.first == "crush_device_class") {
    device_class = param.second;
  } else {
    ss << "unknown param " << param;
    return -EINVAL;
  }
}
```

**Comment by liewegas** [comment]
I'm not sure if it's better to error out on a json arg we don't understand or to just ignore it.  I have no real preference either way. @jcsp?

**Comment by jcsp** [comment]
In general we should ignore unknown fields, but if the struct consists only of unknown fields then we should perhaps error out if there's nothing useful we can do.

**Comment by liewegas** [comment]
an empty struct is technically valid, so that should probably be the same as a dict full of stuff we don't understand.

**Comment by alfredodeza** [comment]
We merged the ceph-volume (pr #20239 and pr #19949) based on the assumption that this would get merged. What can we do to speed the addition of this PR to minimize the amount of time ceph-volume will have a broken flag?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-02-08_20:24:10-rados-wip-sage4-testing-2018-02-08-1235-distro-basic-smithi/

**Comment by liewegas** [comment]
@tchaikov i think it's good to merge now.

**Comment by liewegas** [comment]
backport is https://github.com/ceph/ceph/pull/20405

**Pull request merged**